### ISSUE TITLE
Release cycle plans

### ISSUE THREAD

**Issue by arrdem** [jaunt-lang/jaunt#39]
Right now, `^:deprecated` stuff sticks around indefinitely. It'd be a good idea to put some sort of lifecycle limit on `^:deprecated` vars being removed and stricter guards being enforced. This begs the question of what the release cycle goals are for this project. Since it's my personal gig, I could do daily releases with daily breaking changes and have no problems, but that's pretty user hostile.

I think this is really a question of exactly how unstable Clojarr is. Is Clojarr supposed to be a drop in replacement for Clojure? That isn't going to happen if only due to ... creative differences.

I'm just having this thought as of 6594cf97c500e7e6d7a97b094cb3acad940a196b, which is a binary compatibility breaking change.

**Comment by arrdem** [comment]
Proposed PR policy:
- Adopt @bbatsov's "squash then merge" workflow.
- PRs must test cleanly (already in force)
- PRs must whitespace lint cleanly (already in force for Java #16, need clj support #31)
- PRs must include a changelog entry (in force as of #84)

Proposed release policy:
- Version starts at `1.9.0-$SHASUM-SNAPSHOT`
- First release will be `1.9.0` whenever #69 is closed
- SemVer applies to all present and ongoing work
- Goal of continuous patch releases (automatable?)
- Goal of a monthly minor version release and blog post
- No timeline for major version releases
- `^:deprecated` symbols and namespaces will be _removed_ on major version releases

**Comment by danielcompton** [comment]
What about matching major releases with Clojure's releases? It makes some things simpler, and others more complex.

**Comment by arrdem** [comment]
I don't expect that Clojure will ever see `2.0.0` which would be the next major version. If I had my way, we'd start at version `0.0.0` or `1.0.0` and run from there since I consider whatever this project may be to be fully divorced from Clojure. Unfortunately existing Clojure libraries look at `clojure.core/*clojure-version*` to determine how to behave. Going to an "older" version without fixing `*clojure-version*` to be valued something... else will break stuff including CIDER.

An option is to say that there shall exist `*$NAME-version*` in addition to `*clojure-version*` and that `*clojure-version*` shall be pegged at the "compatible" Clojure version (presumably `1.8.0` for a while to come) and that the other version will track the version of this project.

**Comment by turbopape** [comment]
Regarding this, there is a new Github feature, Pull Request Templates ,
that can among other things show CheckLists of what a contributor shall do
before submitting her PR:

https://github.com/blog/2111-issue-and-pull-request-templates

2016-02-24 1:19 GMT+00:00 Reid D McKenzie notifications@github.com:

**Comment by andrewhr** [comment]
I would dare to say that the most elegant solution about the internal name is to go with `*jaunt-version*`. In the Clojure style, it _decomplects_ release agendas between both projects. Having a so close version like `1.9.0` (which we can expect will be used this year by Clojure) could bring issues like JRuby folks had with Ruby `1.9.x`/`2.x`. That's why they opted by `9.0.0.0` scheme for the current major release.

**Comment by arrdem** [comment]
Hum. I think that's a reasonable argument, and I agree with the concerns about version number reuse. I'd definitely accept `^:deprecated` flagging `*clojure-version*`, hardcoding it and creating a `*jaunt-version*` which makes use of the mvn git plugin's generated output.

Where should Jaunt start counting is I think the next question. `0.1.0`? `1.0.0`? It's "just a number" but I think that the implications in terms of `0.n.m` being "not there yet" vs `1.n.m` being "ready to go" could be significant.

**Comment by andrewhr** [comment]
Humm... `1.0.0` should be fine given the fact that it's still "just Clojure".

To be more concrete: some improvements like having better warnings and errors are "safe enough". Even the reload stuff I think it's good to go since it should affects more dev environments. If you desire to try more substantial changes like [porting Skummet ideas](https://github.com/jaunt-lang/jaunt/issues/73) could make the things... a bit more adventurous.

**Comment by arrdem** [comment]
One concern with reusing `1.0.0` is that a bunch of Clojure has `^:added` metadata which says `1.0`, `1.1`, `1.2` etc. It doesn't really make a lot of sense to have a new `1.0` when there are a bunch of references to "future" versions kicking around. Maybe that's an argument for going straight for `2.0.0`, or for simply reworking all the `^:added` metadata to make clear which project. I'm honestly not a great fan of that solution, but it would be possible.

**Comment by andrewhr** [comment]
Yeah, I didn't though about that before since it's more difficult to appear on application code which I commonly work on. Defining new semantics for usage of `^:added` is workable, but I'm not a fan either since this kind of improvement adds weight along the time.

As far as you known, how this metadata is consumed in the wild and frequent (you think) is that? How Jaunt should behave relating this version marks is still nebulous, since neither suggested approach sounds "correct" to me.

**Comment by arrdem** [comment]
As with so many things in Clojure, `^:added` is strictly advisory and doesn't do anything. It's just used to record project history. I know of no tooling which makes use of that particular information except for Grimoire and ClojureDocs, which provide special display handling for it and presume it's just an arbitrary `String`. Some gh "grepping" here would probably be informative. Eastwood and other code analysis tools are likely candidates. As you say, for application level code it really doesn't matter.

**Comment by andrewhr** [comment]
I asked because just _maybe_ working on a more flexible mechanism could be desirable, e.g., using a similar metadata mechanism but to advise about the current library - here we have `core` but the usage could be extended towards any library. But sorry the divergence, this of course is out of scope from this issue.

A very superficial grep on Eastwood didn't show any _consumption_ of this metadata, but looks like `contrib` libraries (which are copied inside the tool) uses `:added` on theirs on context.

**Comment by andrewhr** [comment]
How messy would be just `s/ANY_CLOJURE_VERSION/JAUNT_FIRST_VERSION/`? If they point to "current library version" after all... ¯_(ツ)_/¯

**Comment by arrdem** [comment]
The only piece of library code I know of which is cause for concern with regards to version information is this ... beauty. https://github.com/clojure/tools.reader/blob/7039837e121ff7d60b7e6e324a995bcca87f82b4/src/main/clojure/clojure/tools/reader/impl/utils.clj

CIDER and other tooling depends on tools.reader, so maintaining `*clojure-version*` at `a.b.c` for `a=1`,  `b>=7` is a priority in the interests of tooling compatibility. This is easy and already provided for in #95.

It is certainly an option to say "Hokay new project new 1.0" and just mark everything `^:added "1.0.0"` which would I think be the most sane of the various options on the table. I'm a little reluctant over changeset size, but that `^:added` only gives two digits in core is ... odd and arguably worth changing.

**Comment by arrdem** [comment]
Okay. So after thinking about this some more, I think that the most reasonable thing to do is to adopt `0.1.0` as the first "release" in true Clojure community style SemVer be damned, and to update `^:added` to simply state `0.1.0` which with respect to this project is strictly true.

The only complication with this approach is that `1.9.0-RC{1..4}` are deployed [on clojars](https://clojars.org/org.jaunt-lang/jaunt/versions) and heaven willing we'll want to reuse those version numbers some day so getting @danielcompton or someone to purge those (unused) jars is likely in order. Since the project as a whole has 15 downloads, the overwhelming majority of which I'm sure are me, this shouldn't be a problem.

**Comment by arrdem** [comment]
```
# depends on #72
echo "0.1.0" > VERSION
find src test -type f -name "*.clj" -exec sed -i -e 's/:added\( *\)"1.[0-9]"/:added\1"0.1.0"/g' {} \;
lein cljfmt fix
git add src test VERSION
git commit -m "Reset version to 0.1.0"
```
